rfcs/text/0428-extending-types.md

```diff
@@ -0,0 +1,677 @@
+
+```json
+{
+  "allOf": [
+    {
+      "type": "object",
+      "properties": {
+        "city": { "type": "string" }
+      },
+      "required": ["city"],
+      "unevaluatedProperties": false
+    }
+  ],
+  "type": "object",
+  "properties": {
+    "type": { "name": "string" }
+  },
+  "unevaluatedProperties": false
+}
+```
+
+and trying to validate against
+
+```json
+{
+  "city": "Copenhagen",
+  "name": "Charles"
+}
+```
+
+results in validations errors such as `Property 'name' has not been successfully evaluated and the schema does not allow unevaluated properties`.
+
+Changing from `unevaluatedProperties` to `additionalProperties` results in errors `Property 'name' has not been defined and the schema does not allow additional properties`.
+
+Both of these solutions for strict schemas would not be suitable for the type of expressiveness we want for type extension, unfortunately.
+
+The behavior we're after is that `unevaluatedProperties` should only validate at the top level of a type. This would allow supertypes to still validate `unevaluatedProperties`, but defer checking if they're used within in subtype.
+
+### Alternative through open/closed schemas
+
+A proposed way to deal with conditional `unevaluatedProperties` is to use custom `$defs` definitions for open and closed variations of entity types.
+**Open by default schema**:
+
+```json
+{
+  "$id": "https://example.com/schema",
+
+  // ... schema contents  ...
+
+  "$defs": {
+    "closed": {
+      "$anchor": "closed",
+      "$ref": "#",
+      "unevaluatedProperties": false
+    }
+  }
+}
+```
+
+Here, referencing `https://example.com/schema` in a `$ref` will result in an _open schema_ that _does not_ specify `{ "unevaluatedProperties": false }`. Referencing `https://example.com/schema#closed` results in retrieving a _closed schema_ that _does_ specify `{ "unevaluatedProperties": false }`.
+
+**Closed by default schema**:
+
+```json
+{
+  "$id": "https://example.com/schema",
+  "$ref": "#open",
+  "unevaluatedProperties": false,
+
+  "$defs": {
+    "schema": {
+      "$anchor": "open"
+      // ... schema contents ...
+    }
+  }
+}
+```
+
+Here, referencing `https://example.com/schema` in a `$ref` will result in a _closed schema_ that _does_ specify `{ "unevaluatedProperties": false }`. Referencing `https://example.com/schema#open` results in retrieving an _open schema_ that _does not_ specify `{ "unevaluatedProperties": false }`.
+
+(Thanks Jason Desrosiers for the suggestions!)
+
+Using the above setup would mean that we need to specify `#open` at the end of type URIs when URIs appear in `allOf`. We would also need to serve type schemas with one of the above structures, such that they conform with JSON Schema. Although this does add some ceremony around extending types, we would end up with a type extension system that would conform to JSON Schema without implicitness or redefining keywords.
```

Comment:
Do we have a justification for why we aren't going with this approach? (From the wording of the doc it seems like we're not anyway)